javascript-client/src/client.ts

```diff
@@ -49,7 +55,41 @@ export type S3FileFieldProgressCallback = (progress: S3FileFieldProgress) => voi
 export interface S3FileFieldClientOptions {
   readonly baseUrl: string;
   readonly onProgress?: S3FileFieldProgressCallback;
-  readonly apiConfig?: AxiosRequestConfig
+  readonly apiConfig?: AxiosRequestConfig;
+}
+
+function sleep(ms: number): Promise<void> {
+  return new Promise((resolve) => {
+    window.setTimeout(() => { resolve(); }, ms);
+  });
+}
+
+function shouldRetry(error: Error): boolean {
+  // We only retry requests under certain failure modes. Namely, either
+  // network errors, or a subset of HTTP error codes.
+  const axiosErr = (error as AxiosError);
+  return axiosErr.isAxiosError && (
+    !axiosErr.response
+    || [429, 500, 502, 503, 504].includes(axiosErr.response.status)
+  );
+}
+
+async function retry<T>(
```

Comment:
What do you think about using one of these:
* https://www.npmjs.com/package/axios-retry
* https://www.npmjs.com/package/retry-axios

I prefer https://www.npmjs.com/package/retry-axios , since it uses proper Axios interceptors.

Reply:
Personally I'd prefer not to have to rewrite this, but if you think using one of those libraries is going to be a better experience, go ahead and make the decision.

Reply:
Using npmjs.com/package/retry-axios has the potential to remove most of the code (and maintenance burden) here, I think it would be worth trying. @zachmullen If you don't have time, I'd be happy to try adding it.

Reply:
FWIW, I'm using axios-retry in NLI, and it works fine.  retry-axios does look a little more refined though.

Reply:
@brianhelba that would be great if you want to take a crack at it. 👍

Reply:
Not 100% sure, but from the README it looks like `axios-retry` doesn't have a callback when retry is happening, but `retry-axios` does.